### PR TITLE
added realm and basic auth for provider setup, made realm_id and realm optional for resources

### LINKED ISSUE #62: provider should support using other realms for authentication
Currently, this provider assumes you will be using the `master` realm for authentication and configuring Keycloak, but I think it makes sense to want to use other realms for this too, perhaps to limit the scope of a service account to only one realm, or to allow different teams to configure their own realms without being given access to `master`.

In particular, I think the provider would have to support:

- Specifying the realm in config
- Specifying the realm via environment variable (which takes precedence if both are specified?)
- ~If the realm is specified either way, imports should not require a realm ID~
- ~If the realm is specified either way, `realm_id` should not be a required attribute on all resources~

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
Fixes #62

### CODE REVIEW

File: provider/provider.go [hunk 1]
```diff
@@ -34,9 +34,32 @@ func KeycloakProvider() *schema.Provider {
...
+				Optional:      true,
+				Type:          schema.TypeString,
+				DefaultFunc:   schema.EnvDefaultFunc("KEYCLOAK_USERNAME", "admin"),
+				ConflictsWith: []string{"client_secret"},
+			},
+			"password": {
+				Optional:      true,
+				Type:          schema.TypeString,
+				DefaultFunc:   schema.EnvDefaultFunc("KEYCLOAK_PASSWORD", nil),
+				ConflictsWith: []string{"client_secret"},
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply provider/provider.go]: It should not, for password grant type when using public access type client_id is required and for confidential access type client_secret is required too

[the remark this answers is not in the document]
Comment (mrparkers) [reply provider/provider.go]: Ah yes, you're right.  Just curious, do you use the `admin-cli` client for this?  Or do you create your own client with only the access you need?  I'm trying to understand this use case better.

[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply provider/provider.go]: using admin-cli for initial setup of realms and then going to provide an access for different teams to be able to manage only their own realms with their own resources

File: keycloak/keycloak_client.go [hunk 2]
```diff
@@ -147,6 +194,7 @@ func (keycloakClient *KeycloakClient) addRequestHeaders(request *http.Request) {
 
 	if request.Method == http.MethodPost || request.Method == http.MethodPut {
 		request.Header.Set("Content-type", "application/json")
+		request.ContentLength = 0
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/keycloak_client.go]: Already removed it

**Comment by mrparkers** [comment]
Okay @AndrewChubatiuk, I've given it some thought, and I've decided that it would probably be better if adding a `realm` attribute to the provider did not affect resources at all.  All resources (except `keycloak_realm`) should instead explicitly set a `realm_id` to avoid any confusion.

Would you be able to remove the commits that allow resources to assume a default realm from the provider, and add some tests for the keycloak client to ensure a JWT was provisioned for a different realm when specified?

**Comment by AndrewChubatiuk** [comment]
@mrparkers 
Don't think that it makes sense. 
E.g.: AWS terraform provider requires using a separate provider for each AWS region though it's not difficult to add region attribute to every resource that it manages, but it's doesn't make sense. 
The same is here, instead of specifying realm for every resource you can have several providers for each realm and then specify provider alias for each module/resource that you want to create

**Comment by AndrewChubatiuk** [comment]
@mrparkers 
Closing this PR, created #88 instead

**Pull request closed without merging**